src/xunit.performance.api/ETWProfiler.cs

```diff
@@ -0,0 +1,300 @@
+
+                foreach (var cpuInfo in providers.OfType<CpuCounterInfo>())
+                {
+                    if (availableCpuCounters.TryGetValue(cpuInfo.CounterName, out var profInfo))
+                    {
+                        profileSourceIDs.Add(profInfo.ID);
+                        profileSourceIntervals.Add(Math.Min(profInfo.MaxInterval, Math.Max(profInfo.MinInterval, cpuInfo.Interval)));
+                    }
+                }
+
+                if (profileSourceIDs.Count > 0)
+                    TraceEventProfileSources.Set(profileSourceIDs.ToArray(), profileSourceIntervals.ToArray());
+            }
+        }
+
+        private static bool NeedSeparateKernelSession(KernelProviderInfo kernelProviderInfo)
+        {
+            if (kernelProviderInfo == null)
+                return false;
+
+            // Prior to Windows 8 (NT 6.2), all kernel events needed the special kernel session.
+            if (!IsWindows8OrGreater)
+                return true;
+
+            // CPU counters need the special kernel session
+            if (((KernelTraceEventParser.Keywords)kernelProviderInfo.Keywords & KernelTraceEventParser.Keywords.PMCProfile) != 0)
+                return true;
+
+            return false;
+        }
+
+        private static bool IsWindows8OrGreater => IsWindows8OrGreater();
+
+        private static IEnumerable<ProviderInfo> RequiredProviders { get; }
+
+        private static Guid EventSourceGuidFromName { get; }
+
+        [Conditional("DEBUG")]
+        private static void PrintProfilingInformation(string assemblyFileName, string sessionName, string userFullFileName)
+        {
+            WriteDebugLine("  ===== ETW Profiling information =====");
+            WriteDebugLine($"       Assembly: {assemblyFileName}");
+            WriteDebugLine($"   Session name: {sessionName}");
+            WriteDebugLine($"  ETW file name: {userFullFileName}");
+            WriteDebugLine("  =====================================");
+        }
+
+        [Conditional("DEBUG")]
+        private static void PrintAvailableProfileSources()
+        {
+            var availableProfileSources = TraceEventProfileSources.GetInfo();
+            var cpuCounterIds = new List<int>();
+            var cpuCounterIntervals = new List<int>();
+            var sb = new StringBuilder();
+
+            foreach (var kvp in availableProfileSources)
+            {
+                sb.AppendLine();
+                sb.AppendLine($"Profile name: {kvp.Key}");
+                sb.AppendLine($"  ID :          {kvp.Value.ID}");
+                sb.AppendLine($"  Interval :    {kvp.Value.Interval}");
+                sb.AppendLine($"  MaxInterval : {kvp.Value.MaxInterval}");
+                sb.AppendLine($"  MinInterval : {kvp.Value.MinInterval}");
+                sb.AppendLine();
+            }
+
+            WriteDebugLine(sb.ToString());
+        }
+
+        private sealed class PerformanceTestMessage
```

Comment:
@brianrob This is just a port of the xunit runner, and it's used in the class below (Lines 229, 232, 253). I can probably do some cleanup here.

---
In reply to: [100344105](https://github.com/Microsoft/xunit-performance/pull/183#discussion_r100344105) [](ancestors = 100344105)